### ISSUE TITLE
[BUG] - Cloudflare DNS Validation Error

### ISSUE THREAD

**Issue by eliasdabbas** [nebari-dev/nebari-docs#148]
### Describe the bug

* Setting up Nebari on GCP went smoothly in all steps, except for the DNS part on Cloudflare. Failure happens while attempting to create A record on Cloudflare:

```bash
INFO:qhub.provider.dns.cloudflare:record name= type=A address=35.223.225.228 does not exists creating

# Full traceback shared below

CloudFlare.exceptions.CloudFlareAPIError: DNS Validation Error
```

* My domain's name servers have been assigned to Cloudflare already
* All authentication has been setup properly, no issues arose
* Command to setup Nebari: 
```bash
qhub deploy -c qhub-config.yaml --dns-provider cloudflare --dns-auto-provision
```

### Expected behavior

* The DNS provisioning should happen without issues
* The hub should start running online

### OS system and architecture in which you are running Nebari

Mac OS Big Sur v11.4

### How to Reproduce the problem?

* All steps in the documentation were followed (creating a GCP project, credentials, Cloudflare credentials, etc.)
* CLI commands:
```bash
pip install qhub
qhub init gcp
# enable various APIs
qhub deploy -c qhub-config.yaml --dns-provider cloudflare --dns-auto-provision
```

### Command output

```bash session
[terraform]: load_balancer_address = {
[terraform]:   "hostname" = ""
[terraform]:   "ip" = "35.223.225.228"
[terraform]: }

INFO:qhub.provider.terraform:terraform apply took 7.520 [s]
INFO:qhub.provider.terraform:terraform=/var/folders/44/5zl0h4bn07j2q07_dzd_wzdm0000gn/T/terraform/1.0.5/terraform output directory=stages/04-kubernetes-ingress

INFO:qhub.provider.terraform:terraform output took 2.712 [s]

Attempt 1 succeeded to connect to tcp://35.223.225.228:80

Attempt 1 succeeded to connect to tcp://35.223.225.228:8786

Attempt 1 succeeded to connect to tcp://35.223.225.228:8022

Attempt 1 succeeded to connect to tcp://35.223.225.228:8023

Attempt 1 succeeded to connect to tcp://35.223.225.228:9080

Attempt 1 succeeded to connect to tcp://35.223.225.228:443

After stage directory=stages/04-kubernetes-ingress kubernetes ingress available on tcp ports={80, 8786, 8022, 8023, 9080, 443}

INFO:qhub.provider.dns.cloudflare:record name= type=A address=35.223.225.228 does not exists creating

Traceback (most recent call last):

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/bin/qhub", line 8, in <module>

    sys.exit(main())

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/__main__.py", line 7, in main

    cli(sys.argv[1:])

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/cli/__init__.py", line 51, in cli

    args.func(args)

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/cli/deploy.py", line 65, in handle_deploy

    deploy_configuration(

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/deploy.py", line 296, in deploy_configuration

    guided_install(

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/deploy.py", line 217, in guided_install

    provision_ingress_dns(

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/deploy.py", line 115, in provision_ingress_dns

    update_record(zone_name, record_name, "A", ip_or_hostname)

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/qhub/provider/dns/cloudflare.py", line 46, in update_record

    cf.zones.dns_records.post(zone_id, data=record)

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/CloudFlare/cloudflare.py", line 675, in post

    return self._base.do_auth('POST', self._parts, [identifier1, identifier2, identifier3, identifier4], params, data, files)

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/CloudFlare/cloudflare.py", line 128, in do_auth

    return self._call(method, headers, parts, identifiers, params, data, files)

  File "/Users/me/Desktop/temp/nebaritest/nebari-env/lib/python3.10/site-packages/CloudFlare/cloudflare.py", line 497, in _call

    raise CloudFlareAPIError(code, message, error_chain)

CloudFlare.exceptions.CloudFlareAPIError: DNS Validation Error
```

### Versions and dependencies used.

* pip==21.3.1
* qhub==0.4.4
* kubectl: command not found

### Compute environment

GCP

### Anything else?

I hope my description is clear enough. Happy to provide any further required details if needed. 

Thanks!

**Comment by viniciusdc** [comment]
Hi @eliasdabbas, thanks for raising the issue. Regarding the deployment logs above, based on the logs, I don't think this is the issue because we would see another message, but could you validate if the token was set in the env vars? `CLOUDFLARE_TOKEN. '

Regarding the error message, for the DNS auto-provision to work, you will need to:
- Add a valid domain name in the configuration. I think this is not the issue, but could you check if there are any strange characters in the domain name? 
- Also, if you set the domain as `mydomain.test.com,` the root domain `test.com` needs to exist in cloud-flare.

**Comment by eliasdabbas** [comment]
> could you validate if the token was set in the env vars? CLOUDFLARE_TOKEN.

I did, also checked that it's in `~/.bash_profile` , sourced it, tried again, still getting the exact same error.

> could you check if there are any strange characters in the domain name? you can see it by opening the qhub-config.yaml

The domain name is `advertools.app`. I don't think there are issues with that. Yes, searching for the domain name with grep I get 
`domain: advertools.app`

No matter what combination of domain, DNS A record combination I try, I'm getting the exact same error message every time. De/re-activated my venv, removed and added DNS records. Still the same thing. 
Do you think I might be doing something wrong with DNS setup?

More interesting (not so pleasant) data: I already got charged by Google!

Appreciate your help. Thanks.

**Comment by viniciusdc** [comment]
Uhm, that's weird. To make it work for you, just create the domain manually:
- Remove the `--disable-prompt` flag from the deploy command. -- This will give you the freedom to grab the LB IP and update the DNS
- Create the domain DNS record on Cloudflare and point to the IP you received. Once that's done, the deployment shall proceed. Also, remove the `--dns-provider` and `--dns-autoprovision` flags when using the approach; there might be some kind of error with the Cloudflare API: [here are some causes that I could find](https://community.cloudflare.com/t/community-tip-fixing-code-1004-dns-validation-error/46697)

**Comment by viniciusdc** [comment]
Also, @eliasdabbas are you testing out Nebari, or will you have this in a production environment?
Usually, for trying out Nebari, we recommend using the local installation process instead of a cloud deployment,[ here](https://deploy-preview-113--nebari-docs.netlify.app/tutorials/nebari-local-development) are some docs (still in PR, as we are working on fixing a compatibility issue with Mac)

**Comment by eliasdabbas** [comment]
> are you testing out Nebari, or will you have this in a production environment?

Both. This is a testing domain/server, and eventually I want to run it in production. 

> we recommend using the local installation process instead of a cloud deployment

Thanks. I'll give it a try. Yet I want to try out the full thing with a cloud provider and see how to run it properly. 

> Create the domain DNS record on Cloudflare and point to the IP you received.

I'm trying this now. I removed previous DNS records, and running with a simple `qhub deploy -c qhub-config.yaml`. I'm adding an A record for the domain, pointing to the IP address given for the load balancer. Seems it hasn't propagated yet, so i'll keep trying. I'll let you know how it goes.

**Comment by pavithraes** [comment]
>  Seems it hasn't propagated yet, so i'll keep trying. I'll let you know how it goes.

Hi, @eliasdabbas! Just checking in, did it work?

**Comment by eliasdabbas** [comment]
Thanks a lot @pavithraes !

Still propagating :( Sorry for the basic question, but is there a way to speed this up?

**Comment by pavithraes** [comment]
@eliasdabbas Thanks for sharing!

> but is there a way to speed this up?

I don't know of a way, but maybe other folks here have thoughts. :)

**Comment by eliasdabbas** [comment]
I finally got this to run. I don't have a clear reason why it wasn't working, but a few things helped: 

Special thanks to @iameskild and @kcpevey  for their help on this!

- In Cloudflare, we changed the "Proxy status" in the DNS settings from "proxied" to "DNS only"
- I had the another different domain pointing to the same IP address, which might have caused the issue. Sorry my bad!
- Modify the config file as follows: 
```
certificate:
  type: lets-encrypt
  acme_email: <your-email-address>
  acme_server: https://acme-v02.api.letsencrypt.org/directory
```
- I'm doing this in the transition phase from "qhub" to "nebari", so I might have mixed some settings. Now the official nebari release is out, and I have it running and looks good.
